src/EnergyPlus/WindowManagerExteriorThermal.cc

```diff
@@ -0,0 +1,663 @@
+      }
+
+    }
+
+    /////////////////////////////////////////////////////////////////////////////////////////
+    //  CWCEHeatTransferFactory
+    /////////////////////////////////////////////////////////////////////////////////////////
+
+    CWCEHeatTransferFactory::CWCEHeatTransferFactory( SurfaceData const & surface, int const t_SurfNum  ) : 
+      m_Surface( surface ), m_SurfNum( t_SurfNum ), m_SolidLayerIndex( 0 ), m_InteriorBSDFShade( false ), 
+      m_ExteriorShade( false ) {
+      m_Window = SurfaceWindow( t_SurfNum );
+      auto ShadeFlag = m_Window.ShadingFlag;
+
+      m_ConstructionNumber = m_Surface.Construction;
+      m_ShadePosition = ShadePosition::NoShade;
+
+      if( ShadeFlag == IntShadeOn || ShadeFlag == ExtShadeOn || ShadeFlag == IntBlindOn ||
+        ShadeFlag == ExtBlindOn || ShadeFlag == BGShadeOn || ShadeFlag == BGBlindOn ||
+        ShadeFlag == ExtScreenOn ) {
+        m_ConstructionNumber = m_Surface.ShadedConstruction;
+        if( m_Window.StormWinFlag > 0 ) m_ConstructionNumber = m_Surface.StormWinShadedConstruction;
+      }
+
+      m_TotLay = getNumOfLayers();
+
+      if( ShadeFlag == IntShadeOn || ShadeFlag == IntBlindOn ) {
+        m_ShadePosition = ShadePosition::Interior;
+      }
+
+      if( ShadeFlag == ExtShadeOn || ShadeFlag == ExtBlindOn || ShadeFlag == ExtScreenOn ) {
+        m_ShadePosition = ShadePosition::Exterior;
+      }
+
+      if( ShadeFlag == BGShadeOn || ShadeFlag == BGBlindOn ) {
+        m_ShadePosition = ShadePosition::Between;
+      }
+
+    }
+
+    /////////////////////////////////////////////////////////////////////////////////////////
+    shared_ptr< CSingleSystem > CWCEHeatTransferFactory::getTarcogSystem( double const t_HextConvCoeff ) {
+      auto Indoor = getIndoor();
+      auto Outdoor = getOutdoor( t_HextConvCoeff );
+      auto aIGU = getIGU();
+
+      // pick-up all layers and put them in IGU (this includes gap layers as well)
+      for( auto i = 0; i < m_TotLay; ++i ) {
+        auto aLayer = getIGULayer( i + 1 );
+        assert( aLayer != nullptr );
+        // IDF for "standard" windows do not insert gas between glass and shade. Tarcog needs that gas
+        // and it will be created here
+        if( m_ShadePosition == ShadePosition::Interior && i == m_TotLay - 1 ) {
+          auto aAirLayer = getShadeToGlassLayer( i + 1 );
+          aIGU->addLayer( aAirLayer );
+        }
+        aIGU->addLayer( aLayer );
+        if( m_ShadePosition == ShadePosition::Exterior && i == 0 ) {
+          auto aAirLayer = getShadeToGlassLayer( i + 1 );
+          aIGU->addLayer( aAirLayer );
+        }
+      }
+
+      auto aSystem = make_shared< CSingleSystem >( aIGU, Indoor, Outdoor );
+
+      return aSystem;
+    }
+
+    /////////////////////////////////////////////////////////////////////////////////////////
+    MaterialProperties* CWCEHeatTransferFactory::getLayerMaterial( int const t_Index ) const {
```

Comment:
You use `shared_ptr` everywhere else but here you are returning a raw pointer, I assume because it is pointing to an EnergyPlus data structure. I see it is `private` in the class so i'm less concerned about this.